hal-core/src/mem/page/alloc.rs

```diff
@@ -0,0 +1,571 @@
+                    region = ?region2,
+                    min_size = self.min_size,
+                    "leaking a region smaller than min page size"
+                );
+            }
+        }
+
+        // Update the base virtual address of the heap.
+        let region_vaddr = region.base_addr().as_usize() + self.offset();
+        self.base_vaddr.fetch_min(region_vaddr, AcqRel);
+
+        // ...and actually add the block to a free list.
+        let block = Free::new(region, self.offset());
+        unsafe { self.push_block(block) };
+        Ok(())
+    }
+
+    #[tracing::instrument(skip(self), level = "trace")]
+    unsafe fn push_block(&self, block: ptr::NonNull<Free>) {
+        let block_size = block.as_ref().size();
+        let order = self.order_for_size(block_size);
+        tracing::trace!(block = ?block.as_ref(), block.order = order);
+        let free_lists = self.free_lists.as_ref();
+        if order > free_lists.len() {
+            todo!("(eliza): choppity chop chop down the block!");
+        }
+        free_lists[order].lock().push_front(block);
+        let mut sz = self.heap_size.load(Acquire);
+        while let Err(actual) =
+            // TODO(eliza): if this overflows that's bad news lol...
+            self
+            .heap_size
+            .compare_exchange_weak(sz, sz + block_size, AcqRel, Acquire)
+        {
+            sz = actual;
+        }
+    }
+
+    /// Returns `block`'s buddy, if it is free
+    ///
+    /// The "buddy" of a block is the block from which that block was split off
+    /// to reach its current order, and therefore the block with which it could
+    /// be merged to reach the target order.
+    unsafe fn buddy_for(
+        &self,
+        block: ptr::NonNull<Free>,
+        order: usize,
+    ) -> Option<ptr::NonNull<Free>> {
+        let size = self.size_for_order(order);
+        let base = self.base_vaddr.load(Relaxed);
+
+        if base == core::usize::MAX {
+            // This is a bug.
+            tracing::error!("cannot find buddy block; heap not initialized!");
+            return None;
+        }
+
+        tracing::trace!(
+            heap.base = trace::hex(base),
+            block.addr = ?block,
+            block.order = order,
+            block.size = size,
+            "calculating buddy..."
+        );
+
+        // Find the relative offset of `block` from the base of the heap.
+        let rel_offset = block.as_ptr() as usize - base;
+        let buddy_offset = rel_offset ^ size;
+        let buddy = (base + buddy_offset) as *mut Free;
+        tracing::trace!(
+            block.rel_offset = trace::hex(rel_offset),
+            buddy.offset = trace::hex(buddy_offset),
+            buddy.addr = ?buddy,
+        );
+
+        if core::ptr::eq(buddy as *const _, block.as_ptr() as *const _) {
+            tracing::trace!("buddy block is the same as self");
+            return None;
+        }
+
+        if unsafe { (*buddy).magic == Free::MAGIC } {
```

Comment:
Is this sound? What's stopping us from accidentally putting the free magic value into the right place in the block while it's in use?